### PR TITLE
Mark synchronous wrapper as obsolete

### LINKED ISSUE #2253: [Bug] Remove sync-over-async patterns in 7x.
For example: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/f5d06c36ee622f93996e9b7e3e25b8604140cf16/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs#L1473C1-L1474C1

This is usually inefficient and dangerous when present in hot path as it can lead to threadpool starvation.

[Should I expose synchronous wrappers for asynchronous methods? - .NET Parallel Programming (microsoft.com)](https://devblogs.microsoft.com/pfxteam/should-i-expose-synchronous-wrappers-for-asynchronous-methods/)

[AspNetCoreDiagnosticScenarios/AsyncGuidance.md at master · davidfowl/AspNetCoreDiagnosticScenarios (github.com)](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#warning-sync-over-async)

### PR DESCRIPTION
**Pull request by westin-m**
Fixes #2253

### CODE REVIEW

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 1]
```diff
@@ -1470,6 +1470,7 @@ public override SecurityToken ReadToken(string token)
         /// <param name="token">A 'JSON Web Token' (JWT) in JWS or JWE Compact Serialization Format.</param>
         /// <param name="validationParameters">A <see cref="TokenValidationParameters"/>  required for validation.</param>
         /// <returns>A <see cref="TokenValidationResult"/></returns>
+        [Obsolete("Synchronous wrapper over an asynchronous method")]
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: add the aka.ms link https://aka.ms/IdentityModel/7-breaking-changes

maybe add a little more text to the obsolete message. 
I would add `[Obsolete("better message", false)]` so a warning appears in the IDE, but no build breaking.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 2]
```diff
@@ -1470,6 +1470,7 @@ public override SecurityToken ReadToken(string token)
         /// <param name="token">A 'JSON Web Token' (JWT) in JWS or JWE Compact Serialization Format.</param>
         /// <param name="validationParameters">A <see cref="TokenValidationParameters"/>  required for validation.</param>
         /// <returns>A <see cref="TokenValidationResult"/></returns>
+        [Obsolete("JsonWebTokens.ValidateToken(string, TokenValidationParameters) has been deprecated and wil be removed in a future release. Use JsonWebTokens.ValidateTokenAsync(string, TokenValidationParameters) instead. For more information, see https://aka.ms/identitymodel/breakingchanges", false)]
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]:
```suggestion
        [Obsolete("`JsonWebTokens.ValidateToken(string, TokenValidationParameters)` has been deprecated and will be removed in a future release. Use`JsonWebTokens.ValidateTokenAsync(string, TokenValidationParameters)` instead. For more information, see https://aka.ms/IdentityModel/7-breaking-changes", false)]
```

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 3]
```diff
@@ -1470,6 +1470,7 @@ public override SecurityToken ReadToken(string token)
         /// <param name="token">A 'JSON Web Token' (JWT) in JWS or JWE Compact Serialization Format.</param>
         /// <param name="validationParameters">A <see cref="TokenValidationParameters"/>  required for validation.</param>
         /// <returns>A <see cref="TokenValidationResult"/></returns>
+        [Obsolete("`JsonWebTokens.ValidateToken(string, TokenValidationParameters)` has been deprecated and will be removed in a future release. Use `JsonWebTokens.ValidateTokenAsync(string, TokenValidationParameters)` instead. For more information, see https://aka.ms/https://aka.ms/IdentityModel/7-breaking-changes", false)]
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: your link got mangled

**Comment by jennyf19** [comment]
Can you update the tests which are using the deprecated method? This way we will not get warnings in the build either.

**Review by jennyf19 — APPROVED** [review]
:shipit: Thanks @westin-m

**Review by keegan-caruso — APPROVED** [review]

**Pull request merged**